cmd/ci-operator-configresolver/main.go

```diff
@@ -158,20 +161,101 @@ func resolveAndRespond(registryAgent agents.RegistryAgent, config api.ReleaseBui
+	writeJSON(w, logger, config)
+}
+
+func writeJSON(w http.ResponseWriter, logger *logrus.Entry, data interface{}) {
+	jsonData, err := json.MarshalIndent(data, "", "  ")
 	if err != nil {
-		metrics.RecordError("failed to marshal config", configresolverMetrics.ErrorRate)
+		metrics.RecordError("failed to marshal JSON", configresolverMetrics.ErrorRate)
 		w.WriteHeader(http.StatusInternalServerError)
-		fmt.Fprintf(w, "failed to marshal config to JSON: %v", err)
-		logger.WithError(err).Errorf("failed to marshal config to JSON")
+		fmt.Fprintf(w, "failed to marshal JSON: %v", err)
+		logger.WithError(err).Errorf("failed to marshal JSON")
 		return
 	}
+	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
-	if _, err := w.Write(jsonConfig); err != nil {
+	if _, err := w.Write(jsonData); err != nil {
 		logrus.WithError(err).Error("Failed to write response")
 	}
 }
 
+func writeErrorPage(w http.ResponseWriter, pageErr error, status int) {
+	w.Header().Set("Content-Type", "text/plain;charset=UTF-8")
+	w.WriteHeader(status)
+	fmt.Fprintf(w, "%v\n", pageErr)
+}
+
+func resolveStepRegistry(regAgent agents.RegistryAgent, confAgent agents.ConfigAgent) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		trimmedPath := strings.TrimPrefix(req.URL.Path, req.URL.Host)
+		// remove leading path prefix
+		trimmedPath = strings.TrimPrefix(trimmedPath, "/step-registry/")
+		// remove trailing slash
+		trimmedPath = strings.TrimSuffix(trimmedPath, "/")
+		splitURI := strings.Split(trimmedPath, "/")
+		if len(splitURI) == 1 {
+			switch splitURI[0] {
+			//case "search":
+			//	searchHandler(confAgent, w, req)
+			//case "job":
+			//	jobHandler(regAgent, confAgent, w, req)
+			default:
+				writeErrorPage(w, errors.New("Invalid path"), http.StatusNotImplemented)
+			}
+			return
+		} else if len(splitURI) == 2 {
+			switch splitURI[0] {
+			case "reference":
+				referenceHandler(regAgent, w, req)
+				return
+			//case "chain":
+			//	chainHandler(regAgent, w, req)
+			//	return
+			//case "workflow":
+			//	workflowHandler(regAgent, w, req)
+			//	return
+			default:
+				writeErrorPage(w, fmt.Errorf("Component type %s not found", splitURI[0]), http.StatusNotFound)
+				return
+			}
+		}
+		writeErrorPage(w, errors.New("Invalid path"), http.StatusNotImplemented)
+	}
+}
+
+func referenceHandler(agent agents.RegistryAgent, w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	defer func() { logrus.Infof("rendered in %s", time.Since(start)) }()
+	name := path.Base(req.URL.Path)
+
+	refs, _, _, docs, metadata := agent.GetRegistryComponents()
+	if _, ok := refs[name]; !ok {
+		writeErrorPage(w, fmt.Errorf("Could not find reference `%s`. If you reached this page via a link provided in the logs of a failed test, the failed step may be a literal defined step, which does not exist in the step registry. Please look at the job info page for the failed test instead.", name), http.StatusNotFound)
```

Comment:
This doesn't port because nothing links the config

Reply:
Dropped `If you reached...` with f532531a2 -> e80cd5572.